presto-verifier/src/main/java/com/facebook/presto/verifier/QueryRewriter.java

```diff
@@ -0,0 +1,184 @@
+            throw new QueryRewriteException("Cannot rewrite queries that use post-queries");
+        }
+
+        Statement statement = parser.createStatement(query.getQuery());
+        try (Connection connection = DriverManager.getConnection(gatewayUrl, usernameOverride.orElse(query.getUsername()), passwordOverride.orElse(query.getPassword()))) {
+            trySetConnectionProperties(query, connection);
+            if (statement instanceof CreateTableAsSelect) {
+                return rewriteCreateTableAsSelect(connection, query, (CreateTableAsSelect) statement);
+            }
+        }
+
+        throw new QueryRewriteException("Unsupported query type: " + statement.getClass());
+    }
+
+    private Query rewriteCreateTableAsSelect(Connection connection, Query query, CreateTableAsSelect statement)
+            throws SQLException
+    {
+        List<String> parts = new ArrayList<>();
+        int originalSize = statement.getName().getOriginalParts().size();
+        int prefixSize = rewritePrefix.getOriginalParts().size();
+        if (originalSize > prefixSize) {
+            parts.addAll(statement.getName().getOriginalParts().subList(0, originalSize - prefixSize));
+        }
+        parts.addAll(rewritePrefix.getOriginalParts());
+        parts.set(parts.size() - 1, createTemporaryTableName());
+        QualifiedName temporaryTableName = QualifiedName.of(parts);
+        Statement rewritten = new CreateTableAsSelect(temporaryTableName, statement.getQuery(), statement.isNotExists(), statement.getProperties(), statement.isWithData());
+        String createTableAsSql = formatSql(rewritten);
+        String checksumSql = checksumSql(connection, statement, temporaryTableName);
+        String dropTableSql = dropTableSql(temporaryTableName);
+        return new Query(query.getCatalog(), query.getSchema(), ImmutableList.of(createTableAsSql), checksumSql, ImmutableList.of(dropTableSql), query.getUsername(), query.getPassword(), query.getSessionProperties());
+    }
+
+    private void trySetConnectionProperties(Query query, Connection connection)
+            throws SQLException
+    {
+        // Required for jdbc drivers that do not implement all/some of these functions (eg. impala jdbc driver)
+        // For these drivers, set the database default values in the query database
+        try {
+            connection.setClientInfo("ApplicationName", "verifier-rewrite");
+            connection.setCatalog(catalogOverride.orElse(query.getCatalog()));
+            connection.setSchema(schemaOverride.orElse(query.getSchema()));
+        }
+        catch (SQLClientInfoException ignored) {
+            // Do nothing
+        }
+    }
+
+    private String createTemporaryTableName()
+    {
+        return rewritePrefix.getSuffix() + UUID.randomUUID().toString().replace("-", "");
+    }
+
+    private String checksumSql(Connection connection, CreateTableAsSelect createTableAsSelect, QualifiedName table)
```

Comment:
This method seems to be doing too much, tangles multiple concerns: it extracts the select clause, runs the query to get the result types and creates the new query. It might be cleaner to split it apart or to untangle the loop.